### DISCUSSION TITLE
Change class names and file names

### DISCUSSION THREAD

**Discussion by dingo-d** [rectorphp/rector#6946] [Q&A]
Hi!

I'm playing with Rector on some of my older WordPress projects, where I've used WordPress coding standards. That means that the class has to be named using underscores like `Some_Class_Name` and has to be in the file called `class-some-class-name.php`.

Now, I'd like to change this to fit in PSR-4 autoloading that is

`SomeClassName` would be in the `SomeClassName.php`. The classes are split into one class per file, and they do follow the structure that the namespace follows the file structure like in PSR-4.

Is it possible to target partial class names and files so that they match? Is it possible to get the class names as arguments inside the `configure` method?

**Comment by TomasVotruba** [comment]
Hi, for working with class names and file names, check PSR4 rules: https://github.com/rectorphp/rector-src/tree/main/rules/PSR4/Rector
Or in README: https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#psr4

You'll have to create custom rule, that moves around your files the specific way.

**Reply by dingo-d** [reply]
Ok, I'll give it a go with a custom rule. `NormalizeNamespaceByPSR4ComposerAutoloadRector` works great, but I'll have to change the folder names to be uppercased as well because it will change the namespace like this

```bash
    ---------- begin diff ----------
@@ @@
  * @package Old_Namespace\Admin
  */

-namespace Old_Namespace\Admin;
+namespace NewNamespace\admin;
    ----------- end diff -----------
```

This works better for namespaces in my case

```php
	// Change root namespace.
	$services->set(RenameNamespaceRector::class)
		->configure([
			'Old_Namespace' => 'NewNamespace',
		]);
```

Because my folder structure is ok.

I'll combine this with a custom rule 👍🏼 Thanks!

**Reply by TomasVotruba** [reply]
Good luck!

Such rule might be useful to other WordPress developers. Would you like to share it in the Rector core after?

**Reply by dingo-d** [reply]
Sure, no problem, as soon as I write it I'll share it here. I've never written Rector rules before so it'll take me some time, but I'll check the renaming rule and connect the dots 😄

**Reply by dingo-d** [reply]
I'm having some trouble with understanding why this isn't working so any help will be appreciated. I wrote the rule like this

<details>

```php
<?php

declare (strict_types=1);

namespace Utils\Rector\Rector\Renaming;

use PhpParser\Node;
use PhpParser\Node\Stmt\Class_;
use Rector\Core\Rector\AbstractRector;
use Rector\Renaming\NodeManipulator\ClassRenamer;
use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;

/**
 * Rector rule for renaming class names
 */
final class RenameClassNameRector extends AbstractRector
{
	/**
	 * @readonly
	 * @var ClassRenamer
	 */
	private $classRenamer;

	public function __construct(ClassRenamer $classRenamer)
	{
		$this->classRenamer = $classRenamer;
	}

	/**
	 * @throws \Symplify\RuleDocGenerator\Exception\PoorDocumentationException
	 */
	public function getRuleDefinition(): RuleDefinition
	{
		return new RuleDefinition(
			'Renames the Snake_Case classnames into PascalCase, e.g. Some_Class_Name to SomeClassName', [
			new CodeSample(
				<<<'CODE_SAMPLE'
class Some_Class_Name
{
	public function run()
	{
		$this->something();
	}
}
CODE_SAMPLE
, <<<'CODE_SAMPLE'
class SomeClassName
{
	public function run()
	{
		$this->something();
	}
}
CODE_SAMPLE
			)
		]
		);
	}

	/**
	 * @return array<class-string<Node>>
	 */
	public function getNodeTypes(): array
	{
		return [Class_::class];
	}

	/**
	 * @param \PhpParser\Node\Stmt\Class_ $node
	 */
	public function refactor(Node $node): ?Node
	{
		if ($node instanceof Class_) {
			$oldName = $node->name->toString();
			$names = explode('_', $oldName);

			$capitalizedNameParts = [];
			foreach ($names as $namePart) {
				$capitalizedNameParts[] = ucfirst(strtolower($namePart));
			}

			$newName = implode('', $capitalizedNameParts);

			return $this->classRenamer->renameNode($node, [$oldName => $newName]);
		}

		return null;
	}
}
```
</details>


The main logic in the `refactor` method is just taking the class name, exploding by `_` converting parts to have the first part of the name uppercased, and then using the `classRenamer` like in the `RenameClassRector` rule to rename the node. But that's not working and I'm not sure why.

I think I'm missing some crucial part 😅

**Reply by TomasVotruba** [reply]
Looks good at first sight. The problem can be any many places, it's hard to say.
Could you share reproducible repository with failing test? I'll check it :)

**Reply by dingo-d** [reply]
Really appreciate the help, here's the repo with a minimal example, the rule, and the tests

https://github.com/dingo-d/rector-snake-case-rename-rule

You should be able to just run composer install and with `composer test:run` the test

**Reply by TomasVotruba** [reply]
I'm looking into it... could you give me PR access there, so I don't have to create forks to send PR?

**Reply by dingo-d** [reply]
Sure thing, I've sent you the invite 🙂
